File: backend-java/ase-challenge/src/test/java/com/verto/ase_challenge/controller/ProductControllerTest.java
package com.verto.ase_challenge.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.verto.ase_challenge.dto.ProductRequest;
import com.verto.ase_challenge.dto.ProductResponse;
import com.verto.ase_challenge.dto.StockUpdateRequest;
import com.verto.ase_challenge.exception.InsufficientStockException;
import com.verto.ase_challenge.exception.ProductNotFoundException;
import com.verto.ase_challenge.service.ProductService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import java.util.Arrays;
import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@WebMvcTest(ProductController.class)
class ProductControllerTest {
    
    @Autowired
    private MockMvc mockMvc;
    
    @MockBean
    private ProductService productService;
    
    @Autowired
    private ObjectMapper objectMapper;
    
    private ProductResponse testProductResponse;
    private ProductRequest testProductRequest;
    
    @BeforeEach
    void setUp() {
        testProductResponse = new ProductResponse();
        testProductResponse.setId(1L);
        testProductResponse.setName("Test Product");
        testProductResponse.setDescription("Test Description");
        testProductResponse.setStockQuantity(100);
        testProductResponse.setLowStockThreshold(10);
        testProductResponse.setLowStock(false);
        
        testProductRequest = new ProductRequest();
        testProductRequest.setName("Test Product");
        testProductRequest.setDescription("Test Description");
        testProductRequest.setStockQuantity(100);
        testProductRequest.setLowStockThreshold(10);
    }
    
    @Test
    @DisplayName("Should create product successfully")
    void shouldCreateProductSuccessfully() throws Exception {
        // Given
        when(productService.createProduct(any(ProductRequest.class))).thenReturn(testProductResponse);
        
        // When & Then
        mockMvc.perform(post("/api/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(testProductRequest)))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.id").value(1L))
                .andExpect(jsonPath("$.name").value("Test Product"))
                .andExpect(jsonPath("$.stockQuantity").value(100));
        
        verify(productService).createProduct(any(ProductRequest.class));
    }
    
    @Test
    @DisplayName("Should return validation error for invalid product request")
    void shouldReturnValidationErrorForInvalidProductRequest() throws Exception {
        // Given
        ProductRequest invalidRequest = new ProductRequest();
        invalidRequest.setName(""); // Invalid - blank name
        invalidRequest.setStockQuantity(-1); // Invalid - negative stock
        
        // When & Then
        mockMvc.perform(post("/api/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.status").value(400))
                .andExpect(jsonPath("$.error").value("Validation Failed"));
        
        verify(productService, never()).createProduct(any(ProductRequest.class));
    }
    
    @Test
    @DisplayName("Should get product by ID successfully")
    void shouldGetProductByIdSuccessfully() throws Exception {
        // Given
        when(productService.getProductById(1L)).thenReturn(testProductResponse);
        
        // When & Then
        mockMvc.perform(get("/api/products/1"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(1L))
                .andExpect(jsonPath("$.name").value("Test Product"));
        
        verify(productService).getProductById(1L);
    }
    
    @Test
    @DisplayName("Should return 404 when product not found")
    void shouldReturn404WhenProductNotFound() throws Exception {
        // Given
        when(productService.getProductById(999L)).thenThrow(new ProductNotFoundException("Product not found with ID: 999"));
        
        // When & Then
        mockMvc.perform(get("/api/products/999"))
                .andExpect(status().isNotFound())
                .andExpect(jsonPath("$.status").value(404))
                .andExpect(jsonPath("$.error").value("Product Not Found"));
        
        verify(productService).getProductById(999L);
    }
    
    @Test
    @DisplayName("Should get all products successfully")
    void shouldGetAllProductsSuccessfully() throws Exception {
        // Given
        List<ProductResponse> products = Arrays.asList(testProductResponse);
        when(productService.getAllProducts()).thenReturn(products);
        
        // When & Then
        mockMvc.perform(get("/api/products"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$").isArray())
                .andExpect(jsonPath("$[0].id").value(1L));
        
        verify(productService).getAllProducts();
    }
    
    @Test
    @DisplayName("Should update product successfully")
    void shouldUpdateProductSuccessfully() throws Exception {
        // Given
        when(productService.updateProduct(anyLong(), any(ProductRequest.class))).thenReturn(testProductResponse);
        
        // When & Then
        mockMvc.perform(put("/api/products/1")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(testProductRequest)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(1L));
        
        verify(productService).updateProduct(eq(1L), any(ProductRequest.class));
    }
    
    @Test
    @DisplayName("Should delete product successfully")
    void shouldDeleteProductSuccessfully() throws Exception {
        // Given
        doNothing().when(productService).deleteProduct(1L);
        
        // When & Then
        mockMvc.perform(delete("/api/products/1"))
                .andExpect(status().isNoContent());
        
        verify(productService).deleteProduct(1L);
    }
    
    @Test
    @DisplayName("Should increase stock successfully")
    void shouldIncreaseStockSuccessfully() throws Exception {
        // Given
        StockUpdateRequest request = new StockUpdateRequest(50);
        testProductResponse.setStockQuantity(150);
        when(productService.increaseStock(1L, 50)).thenReturn(testProductResponse);
        
        // When & Then
        mockMvc.perform(post("/api/products/1/stock/increase")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.stockQuantity").value(150));
        
        verify(productService).increaseStock(1L, 50);
    }
    
    @Test
    @DisplayName("Should decrease stock successfully")
    void shouldDecreaseStockSuccessfully() throws Exception {
        // Given
        StockUpdateRequest request = new StockUpdateRequest(30);
        testProductResponse.setStockQuantity(70);
        when(productService.decreaseStock(1L, 30)).thenReturn(testProductResponse);
        
        // When & Then
        mockMvc.perform(post("/api/products/1/stock/decrease")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.stockQuantity").value(70));
        
        verify(productService).decreaseStock(1L, 30);
    }
    
    @Test
    @DisplayName("Should return 400 when insufficient stock for decrease")
    void shouldReturn400WhenInsufficientStockForDecrease() throws Exception {
        // Given
        StockUpdateRequest request = new StockUpdateRequest(150);
        when(productService.decreaseStock(1L, 150))
            .thenThrow(new InsufficientStockException("Insufficient stock. Available: 100, Requested: 150"));
        
        // When & Then
        mockMvc.perform(post("/api/products/1/stock/decrease")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.status").value(400))
                .andExpect(jsonPath("$.error").value("Insufficient Stock"));
        
        verify(productService).decreaseStock(1L, 150);
    }
    
    @Test
    @DisplayName("Should return validation error for invalid stock update request")
    void shouldReturnValidationErrorForInvalidStockUpdateRequest() throws Exception {
        // Given
        StockUpdateRequest invalidRequest = new StockUpdateRequest(0); // Invalid - must be at least 1
        
        // When & Then
        mockMvc.perform(post("/api/products/1/stock/increase")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.status").value(400))
                .andExpect(jsonPath("$.error").value("Validation Failed"));
        
        verify(productService, never()).increaseStock(anyLong(), anyInt());
    }
    
    @Test
    @DisplayName("Should get low stock products successfully")
    void shouldGetLowStockProductsSuccessfully() throws Exception {
        // Given
        ProductResponse lowStockProduct = new ProductResponse();
        lowStockProduct.setId(2L);
        lowStockProduct.setName("Low Stock Product");
        lowStockProduct.setStockQuantity(5);
        lowStockProduct.setLowStockThreshold(10);
        lowStockProduct.setLowStock(true);
        
        List<ProductResponse> lowStockProducts = Arrays.asList(lowStockProduct);
        when(productService.getLowStockProducts()).thenReturn(lowStockProducts);
        
        // When & Then
        mockMvc.perform(get("/api/products/low-stock"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$").isArray())
                .andExpect(jsonPath("$[0].id").value(2L))
                .andExpect(jsonPath("$[0].lowStock").value(true));
        
        verify(productService).getLowStockProducts();
    }
}

File: backend-ts/src/__tests__/ProductController.test.ts
import request from 'supertest';
import { createApp } from '../app';
import { Database } from '../database/Database';

describe('ProductController', () => {
    let app: any;
    let database: Database;

    beforeEach(async () => {
        database = new Database(':memory:');
        await database.waitForInitialization();
        app = createApp(database);
    });

    afterEach(() => {
        database.close();
    });

    describe('GET /api/products', () => {
        it('should return all products', async () => {
            const response = await request(app)
                .get('/api/products')
                .expect(200);

            expect(Array.isArray(response.body)).toBe(true);
            expect(response.body.length).toBeGreaterThan(0);
            expect(response.body[0]).toHaveProperty('id');
            expect(response.body[0]).toHaveProperty('name');
        });
    });

    describe('GET /api/products/:id', () => {
        it('should return a product by ID', async () => {
            const response = await request(app)
                .get('/api/products/1')
                .expect(200);

            expect(response.body).toHaveProperty('id', 1);
            expect(response.body).toHaveProperty('name');
        });

        it('should return 404 for non-existent product', async () => {
            const response = await request(app)
                .get('/api/products/999')
                .expect(404);

            expect(response.body).toHaveProperty('error', 'Product Not Found');
        });

        it('should return 400 for invalid ID', async () => {
            const response = await request(app)
                .get('/api/products/invalid')
                .expect(400);

            expect(response.body).toHaveProperty('error', 'Bad Request');
        });
    });

    describe('POST /api/products', () => {
        it('should create a new product', async () => {
            const productData = {
                name: 'Test Product',
                description: 'Test Description',
                stockQuantity: 100,
                lowStockThreshold: 10
            };

            const response = await request(app)
                .post('/api/products')
                .send(productData)
                .expect(201);

            expect(response.body).toHaveProperty('id');
            expect(response.body.name).toBe('Test Product');
            expect(response.body.stockQuantity).toBe(100);
        });

        it('should return 400 for invalid product data', async () => {
            const productData = {
                name: '',
                stockQuantity: -1
            };

            const response = await request(app)
                .post('/api/products')
                .send(productData)
                .expect(400);

            expect(response.body).toHaveProperty('error', 'Validation Failed');
            expect(response.body).toHaveProperty('validationErrors');
        });
    });

    describe('PUT /api/products/:id', () => {
        it('should update an existing product', async () => {
            const updateData = {
                name: 'Updated Product',
                stockQuantity: 150
            };

            const response = await request(app)
                .put('/api/products/1')
                .send(updateData)
                .expect(200);

            expect(response.body.name).toBe('Updated Product');
            expect(response.body.stockQuantity).toBe(150);
        });

        it('should return 404 for non-existent product', async () => {
            const updateData = { name: 'Updated Product' };

            const response = await request(app)
                .put('/api/products/999')
                .send(updateData)
                .expect(404);

            expect(response.body).toHaveProperty('error', 'Product Not Found');
        });
    });

    describe('DELETE /api/products/:id', () => {
        it('should delete an existing product', async () => {
            await request(app)
                .delete('/api/products/1')
                .expect(204);

            // Verify product is deleted
            await request(app)
                .get('/api/products/1')
                .expect(404);
        });

        it('should return 404 for non-existent product', async () => {
            const response = await request(app)
                .delete('/api/products/999')
                .expect(404);

            expect(response.body).toHaveProperty('error', 'Product Not Found');
        });
    });

    describe('POST /api/products/:id/stock/increase', () => {
        it('should increase stock quantity', async () => {
            const stockUpdate = { quantity: 50 };

            const response = await request(app)
                .post('/api/products/1/stock/increase')
                .send(stockUpdate)
                .expect(200);

            expect(response.body.stockQuantity).toBeGreaterThan(0);
        });

        it('should return 400 for invalid quantity', async () => {
            const stockUpdate = { quantity: 0 };

            const response = await request(app)
                .post('/api/products/1/stock/increase')
                .send(stockUpdate)
                .expect(400);

            expect(response.body).toHaveProperty('error', 'Validation Failed');
        });
    });

    describe('POST /api/products/:id/stock/decrease', () => {
        it('should decrease stock quantity', async () => {
            const stockUpdate = { quantity: 5 };

            const response = await request(app)
                .post('/api/products/1/stock/decrease')
                .send(stockUpdate)
                .expect(200);

            expect(response.body.stockQuantity).toBeGreaterThanOrEqual(0);
        });

        it('should return 400 for insufficient stock', async () => {
            const stockUpdate = { quantity: 1000 };

            const response = await request(app)
                .post('/api/products/1/stock/decrease')
                .send(stockUpdate)
                .expect(400);

            expect(response.body).toHaveProperty('error', 'Insufficient Stock');
        });
    });

    describe('GET /api/products/low-stock', () => {
        it('should return low stock products', async () => {
            const response = await request(app)
                .get('/api/products/low-stock')
                .expect(200);

            expect(Array.isArray(response.body)).toBe(true);

            // All returned products should have lowStock = true
            response.body.forEach((product: any) => {
                expect(product.lowStock).toBe(true);
            });
        });
    });

    describe('GET /health', () => {
        it('should return health status', async () => {
            const response = await request(app)
                .get('/health')
                .expect(200);

            expect(response.body).toHaveProperty('status', 'OK');
            expect(response.body).toHaveProperty('service', 'Inventory Management API');
        });
    });
});